Handle T message and prepare for adding -p option

datadius opened the pull request:
In protocol.go I added a new field to the Response struct, indicating the type of Protocol. I believe this will prepare us to handle also -d in the future, but for the moment, the goal is to prepare for -p requests.

When parsing the message, I also make sure that the protocol is being followed, as SCP doesn't allow any other messages besides the ones starting with C, T, D.
![image](https://github.com/bramvdbogaerde/go-scp/assets/16731794/aff84892-45b4-458b-a383-0dc803cff64b)

I also added some new methods to check which type of protocol it is and if it is following the protocol. Of course if we get any directory response, it will throw an error saying it doesn't follow protocol, but that would a problem with a custom server sending this information because we don't have methods requesting that yet.

Finally, I added a method for parsing the time components. The length of the number is always from 0:10, so it should throw an error if that is not the case.


In client.go, I modified the client so when we get time information, then we make sure it doesn't impact the flow. What I encountered with a custom ssh server is that, even though I sent only -f, I would receive the time and that would cause an EOF error, as I was asking for more to read than there was on the server.

In this current form, then the time is stored in an object and it will throw an error only if the custom server sends an malformed T message, but then again, there are issues in the custom ssh server.

I prepared the FilesInfo object to use that later on as output for our -p request. For now it's "dead code" as it's not used, but I didn't want to stuff too much in the pull request.

I have tested on a custom ssh server, with the localhost OpenSSH server and using the tests in the repository.

bramvdbogaerde commented on client.go:
```diff
@@ -328,8 +375,50 @@ func (a *Client) CopyFromRemotePassThru(ctx context.Context, w io.Writer, remote
+
+			fileInfo.Update(timeInfo)
+
+			res, err = ParseResponse(r)
```
I think we would need to send `Ack` first before reading the response, otherwise the remote might not proceed with the next message.

datadius replied:
I haven't implemented the -p functionality yet. I did it like this because on the custom ssh server(implemented by wish) the T and C message come one after another without waiting for an Ack. When you use -p, then you're right, the Ack has to be there. I thought of another fix where I read and process until the buffer is ready.

bramvdbogaerde commented on client.go:
```diff
@@ -328,8 +375,50 @@ func (a *Client) CopyFromRemotePassThru(ctx context.Context, w io.Writer, remote
+
+			if res.NoStandardProtocolType() {
+				errCh <- errors.New(fmt.Sprintf("Input from server doesn't follow protocol: %s", res.GetMessage()))
+				return
```
This error handling code (line 396 through 408) is getting quite long and duplicates most of the logic. Communicating a failure either through the actual `err` value of `ParseResponse` or through a method `Error()` might help to reduce this. I know you inherited most of that error handling code so to me, this is not blocking for this PR in particular, but it would improve the code quality in my opinion.

bramvdbogaerde commented on protocol.go:
```diff
@@ -22,6 +22,13 @@ const (
+type ProtocolType = rune
+
+const (
+	Chmod ProtocolType = 'C'
```
Not sure if it is called a `Chmod` message, might be a confusing term since the message is also used for signaling the size of the file and also the start of its transfer. I would keep it if it is part of some standard terminology, otherwise maybe call it `Create`?

bramvdbogaerde replied:
Or `Copy`.

bramvdbogaerde commented on protocol.go:
```diff
@@ -35,8 +42,9 @@ const (
-	Message string
+	Type         ResponseType
+	Message      string
+	ProtocolType rune
```
What is the reason of using a `rune` type here? A `rune` is 32 bits in size and mostly used for encoding unicode code points. However further down we use it to store a single unsigned byte, so it seems that we are wasting some memory space here.

Moreover, you seem to store the same data in `Type` and `ProtocolType`, is this because you want to differentiate between error messages and protocol messages? Maybe we should simply not store the error status and return it as an actual error from `ParseResponse` (see above).

bramvdbogaerde reviewed (changes requested):
Most of the code looks pretty good, thanks for your contribution. I left some minor comments that you might want to look at.

Regards,

Bram

datadius commented:
I understand if the changes I introduced in this last commit are too big 😅 

As you will see, now the CopyFromRemotePassThru method is a lot cleaner and I ended up moving quite a bit too ParseResponse. Everything is handled through errors instead of the methods previously in Response. I ended up removing Response because it wasn't needed anymore.

datadius commented:
Test failed because I updated the error message
Expected: scp: /input/no_such_file.txt: No such file or directory
Got: Failed to execute command with a warning or error: scp: /input/no_such_file.txt: No such file or directory

Should the test be updated, or should I just remove the extra text @bramvdbogaerde ?

The error comes from line 49 in protocol?

bramvdbogaerde commented:
I would remove the extra text lest someone is pattern matching on original.

datadius commented:
> I would remove the extra text lest someone is pattern matching on original.

The test should pass now, I removed the bit I added. I could work on the tests in another pull request. I will have to update the tests anyways once the -p is in there.

datadius reviewed (commented):
changes are ready. 

EDIT:
Thought it would remove the "changes requested"

bramvdbogaerde commented:
> changes are ready.
> 
> EDIT: Thought it would remove the "changes requested"

Yes I have to approve the changes manually.
